search_log.go

```diff
@@ -136,34 +177,45 @@ func readLine(reader *bufio.Reader) (string, error) {
 	return string(line), nil
 }
 
-func readLastLine(file *os.File) string {
-	var line []byte
-	var cursor int64
-	stat, _ := file.Stat()
-	filesize := stat.Size()
+// Read lines from the end of a file
+// endCursor initial value should be the filesize
+func readLastLines(file *os.File, endCursor int64) string {
+	var lines []byte
+	var cursor = endCursor
 	for {
-		cursor -= 1
-		file.Seek(cursor, io.SeekEnd)
-
-		char := make([]byte, 1)
-		file.Read(char)
-
-		// stop if we find a line
-		if cursor != -1 && (char[0] == 10 || char[0] == 13) {
+		// stop if we are at the begining
+		// check it in the start to avoid read beyond the size
+		if cursor <= 0 {
 			break
 		}
-		line = append(line, char[0])
-		if cursor == -filesize { // stop if we are at the begining
-			break
+
+		var size int64 = 512
+		if cursor < size {
+			size = cursor
+		}
+		cursor -= size
+
+		file.Seek(cursor, io.SeekStart)
+		chars := make([]byte, size)
```

Comment:
We can move this line out of the loop and reuse the buffer to improve performance

Reply:
Good idea! update it soon.

Reply:
I tried but found it doesn't work well, including keep size value not change. Because `file.Read(chars)` reads `len(chars)` bytes content, so we need to make sure the chars is allocated by the size we need to read, instead of fixed 512 bytes.